<a href="https://colab.research.google.com/github/yleessam/2412_ml_env/blob/main/4%EC%9E%A5_%EA%B3%A0%EA%B0%9D%EB%A7%8C%EC%A1%B1%EB%8F%84%EB%B6%84%EB%A5%98%EC%98%88%EC%B8%A1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib
import warnings
warnings.filterwarnings('ignore')

cust_df = pd.read_csv("./santander.csv", encoding='latin-1')
print('dataset shape:', cust_df.shape)
cust_df.head(3)

dataset shape: (76020, 371)


ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                      0.0                      0.0  ...   
1                      0.0                      0.0  ...   
2                      0.0                      0.0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3     var38  TARGET  
0                     0.0                     0.0  39205.17       0  
1                     0.0                     0.0  49278.03       0  
2                     0.0                     0.0  67333.77       0  

[3 rows x 371 columns]

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


산탄데르 은행 고객 만족도조사 결과
피처 - 익명화, 컬럼명 익명화, 수치, 0인 데이터가 대다수 - 희소성
타겟 - TARGET

# 데이터 탐색

In [ ]:
cust_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76020 entries, 0 to 76019
Columns: 371 entries, ID to TARGET
dtypes: float64(111), int64(260)
memory usage: 215.2 MB


100명중 4명이 불만족하는 데이터

In [4]:
cust_df['TARGET'].value_counts() #0 만족, 1 불만족
unsatisfied_cnt = cust_df[cust_df['TARGET']==1].TARGET.count()
total_cnt = cust_df.TARGET.count()

print(f'불만족 비율은 : {unsatisfied_cnt/total_cnt:.2f}')

불만족 비율은 : 0.04


최소값, Q1, Q2, Q3 가 0인 피처가 많다.
최소값 -999999.000000 인 피처 var3 > Nan > 중위값/최빈값 2로 대체

In [ ]:
cust_df.describe()

ID           var3         var15  imp_ent_var16_ult1  \
count   76020.000000   76020.000000  76020.000000        76020.000000   
mean    75964.050723   -1523.199277     33.212865           86.208265   
std     43781.947379   39033.462364     12.956486         1614.757313   
min         1.000000 -999999.000000      5.000000            0.000000   
25%     38104.750000       2.000000     23.000000            0.000000   
50%     76043.000000       2.000000     28.000000            0.000000   
75%    113748.750000       2.000000     40.000000            0.000000   
max    151838.000000     238.000000    105.000000       210000.000000   

       imp_op_var39_comer_ult1  imp_op_var39_comer_ult3  \
count             76020.000000             76020.000000   
mean                 72.363067               119.529632   
std                 339.315831               546.266294   
min                   0.000000                 0.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max               12888.030000             21024.810000   

       imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
count             76020.000000             76020.000000   
mean                  3.559130                 6.472698   
std                  93.155749               153.737066   
min                   0.000000                 0.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max                8237.820000             11073.570000   

       imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
count             76020.000000             76020.000000  ...   
mean                  0.412946                 0.567352  ...   
std                  30.604864                36.513513  ...   
min                   0.000000                 0.000000  ...   
25%                   0.000000                 0.000000  ...   
50%                   0.000000                 0.000000  ...   
75%                   0.000000                 0.000000  ...   
max                6600.000000              6600.000000  ...   

       saldo_medio_var33_hace2  saldo_medio_var33_hace3  \
count             76020.000000             76020.000000   
mean                  7.935824                 1.365146   
std                 455.887218               113.959637   
min                   0.000000                 0.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max               50003.880000             20385.720000   

       saldo_medio_var33_ult1  saldo_medio_var33_ult3  \
count            76020.000000            76020.000000   
mean                12.215580                8.784074   
std                783.207399              538.439211   
min                  0.000000                0.000000   
25%                  0.000000                0.000000   
50%                  0.000000                0.000000   
75%                  0.000000                0.000000   
max             138831.630000            91778.730000   

       saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
count             76020.000000             76020.000000   
mean                 31.505324                 1.858575   
std                2013.125393               147.786584   
min                   0.000000                 0.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max              438329.220000             24650.010000   

       saldo_medio_var44_ult1  saldo_medio_var44_ult3         var38  \
count            76020.000000            76020.000000  7.602000e+04   

In [ ]:
cust_df.var3.value_counts()

 2         74165
 8           138
-999999      116
 9           110
 3           108
           ...  
 231           1
 188           1
 168           1
 135           1
 87            1
Name: var3, Length: 208, dtype: int64

## 피처 선택 - id 삭제, var3 은 최빈값 2로 대체

In [9]:
cust_df.drop('ID', axis=1, inplace=True)
cust_df['var3'].replace(-999999, 2, inplace=True)

cust_df.head()

var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0     2     23                 0.0                      0.0   
1     2     34                 0.0                      0.0   
2     2     23                 0.0                      0.0   
3     2     37                 0.0                    195.0   
4     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  imp_op_var40_ult1  ...  \
0                      0.0                      0.0                0.0  ...   
1                      0.0                      0.0                0.0  ...   
2                      0.0                      0.0                0.0  ...   
3                      0.0                      0.0                0.0  ...   
4                      0.0                      0.0                0.0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   
3                      0.0                      0.0                     0.0   
4                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3          var38  TARGET  
0                     0.0                     0.0   39205.170000       0  
1                     0.0                     0.0   49278.030000       0  
2                     0.0                     0.0   67333.770000       0  
3                     0.0                     0.0   64007.970000       0  
4                     0.0                     0.0  117310.979016       0  

[5 rows x 370 columns]

## X, y 데이터 분할


In [10]:
X_features = cust_df.iloc[:, :-1]
y_labels = cust_df.iloc[:, -1]
X_features.shape, y_labels.shape

((76020, 369), (76020,))

## 훈련/테스트 분할

In [11]:
from sklearn.model_selection import train_test_split

In [12]:
X_train, X_test, y_train, y_test = train_test_split(X_features, y_labels, test_size=0.2, random_state=0)
#X_train.count()
print(f'학습 데이터 shape : {X_train.shape}, 테스트 데이터 shape : {X_test.shape}')

train_cnt = y_train.count()
test_cnt = y_test.count()
print(f'학습 데이터 값의 비율 : {y_train.value_counts()/train_cnt}')
print(f'테스트 데이터  값의 비율 : {y_test.value_counts()/test_cnt}')

학습 데이터 shape : (60816, 369), 테스트 데이터 shape : (15204, 369)
학습 데이터 값의 비율 : TARGET
0    0.960964
1    0.039036
Name: count, dtype: float64
테스트 데이터  값의 비율 : TARGET
0    0.9583
1    0.0417
Name: count, dtype: float64


# DT 예측기

# RF 예측기

# XGBoost 예측기 p271

## 훈련 데이터 > 훈련:검증 분할

In [13]:
X_tr, X_val, y_tr, y_val = train_test_split(X_train, y_train, test_size=0.3, random_state=0)

## XGBoost model

In [16]:
%%time

from xgboost import XGBClassifier
from sklearn.metrics import roc_auc_score

xgb_clf = XGBClassifier(n_estimator=500, learning_rate=0.05, random_state=156, early_stopping_rounds=100, eval_metric='auc')
xgb_clf.fit(X_tr, y_tr, eval_set=[(X_tr, y_tr), (X_val, y_val)] )
#xgb_clf.fit(X_tr, y_tr, eval_metric='auc', eval_set=[(X_tr, y_tr), (X_val, y_val)])
xgb_roc_score = roc_auc_score(y_test, xgb_clf.predict_proba(X_test)[:, 1])
xgb_roc_score

[0]	validation_0-auc:0.83603	validation_1-auc:0.80978
[1]	validation_0-auc:0.83783	validation_1-auc:0.81126
[2]	validation_0-auc:0.83892	validation_1-auc:0.81192
[3]	validation_0-auc:0.84090	validation_1-auc:0.81349
[4]	validation_0-auc:0.84187	validation_1-auc:0.81356
[5]	validation_0-auc:0.84355	validation_1-auc:0.81422
[6]	validation_0-auc:0.84534	validation_1-auc:0.81550
[7]	validation_0-auc:0.84658	validation_1-auc:0.81623
[8]	validation_0-auc:0.84727	validation_1-auc:0.81669
[9]	validation_0-auc:0.84794	validation_1-auc:0.81668
[10]	validation_0-auc:0.84980	validation_1-auc:0.81828
[11]	validation_0-auc:0.85105	validation_1-auc:0.81965
[12]	validation_0-auc:0.85217	validation_1-auc:0.81999
[13]	validation_0-auc:0.85301	validation_1-auc:0.82049
[14]	validation_0-auc:0.85558	validation_1-auc:0.82210
[15]	validation_0-auc:0.85662	validation_1-auc:0.82288
[16]	validation_0-auc:0.85811	validation_1-auc:0.82380
[17]	validation_0-auc:0.85864	validation_1-auc:0.82365
[18]	validation_0-au

0.8413222688684454

# 하이퍼오피트 파라미터 최적화

노트북을 열면 최신버전의 사이킷런이 설치되어 있어요.   
하이퍼오피티, xgbm의 현재 설치버전과 의존성이 일치하는 사이킷런 버전이 1.3.0 이래요.

사이킷런 버전을 다운시킵니다.
!pip install scikit-learn==1.3.0

실행후 코랩 노트북의 런타임 메뉴에서 세션 다시 시작을 선택합니다.
그리고, 데이터 로드부터 다시 합니다.
!pip install scikit-learn==1.3.0 는 실행하지 않습니다.

버전 체크합니다.   
('2.1.3', '0.2.7', '1.3.0')



In [35]:
!pip install scikit-learn==1.3.0

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.8/10.8 MB 45.1 MB/s eta 0:00:00
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.6.0
    Uninstalling scikit-learn-1.6.0:
      Successfully uninstalled scikit-learn-1.6.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
mlxtend 0.23.3 requires scikit-learn>=1.3.1, but you have scikit-learn 1.3.0 which is incompatible.


In [14]:
import xgboost
import hyperopt
import sklearn
xgboost.__version__, hyperopt.__version__, sklearn.__version__

('2.1.3', '0.2.7', '1.3.0')

In [15]:
from hyperopt import fmin, tpe, Trials, STATUS_OK
from sklearn.base import BaseEstimator, ClassifierMixin
from xgboost import XGBClassifier
from sklearn.model_selection import cross_val_score
from hyperopt  import hp

In [16]:
# Sklearn 호환 XGBoost 클래스 정의
class SklearnCompatibleXGBClassifier(XGBClassifier, BaseEstimator, ClassifierMixin):
    def __sklearn_tags__(self):
        return {"non_deterministic": True}

In [17]:
from hyperopt.pyll.base import scope
import numpy as np

# 하이퍼파라미터 검색공간
search_space = { 'n_estimators': scope.int(hp.quniform('n_estimators', 50, 300, 10)),
                'max_depth': scope.int(hp.quniform('max_depth', 3, 10, 1)),
                'learning_rate': hp.loguniform('learning_rate', np.log(0.01), np.log(0.3)),
                'subsample':hp.uniform('subsample', 0.5, 1.0),
                'colsample_bytree': hp.uniform('colsample_bytree', 0.5, 1.0),
                }

In [18]:
from hyperopt import fmin, tpe, Trials, STATUS_OK
from xgboost import XGBClassifier
from sklearn.model_selection import cross_val_score


def objective_func_xgb(params):

    model = SklearnCompatibleXGBClassifier(
        n_estimators=100, #params['n_estimators'],
        max_depth=params['max_depth'],
        learning_rate=params['learning_rate'],
        #subsample=params['subsample'],
        colsample_bytree=params['colsample_bytree'],
        random_state=42,
        eval_metric='logloss'
    )
    #k-fold 추가

    score_mean = cross_val_score(model, X_train, y_train, cv=5, scoring='accuracy').mean()
    return {'loss': -1*score_mean , 'status': STATUS_OK}

In [20]:
trials = Trials()

best_params = fmin(
    fn=objective_func_xgb,
    space=search_space,
    algo=tpe.suggest,
    max_evals=50,
    trials=trials
)

100%|██████████| 50/50 [30:06<00:00, 36.13s/trial, best loss: -0.9610628759849597]


In [22]:
best_params

{'colsample_bytree': 0.5132064441508775,
 'learning_rate': 0.056778045073597835,
 'max_depth': 8.0,
 'n_estimators': 220.0,
 'subsample': 0.6525052716600095}

{'colsample_bytree': 0.5325981930337212,
 'learning_rate': 0.17220434779843125,
 'max_depth': 4.0,
 'n_estimators': 140.0,
 'subsample': 0.7272220988660021}

In [37]:
best_model = XGBClassifier(n_estimators = int(best_params['n_estimators']),
                           max_depth = int(best_params['max_depth']),
                           learning_rate=best_params['learning_rate'],
                           subsample=best_params['subsample'] ,
                           colsample_bytree=best_params['colsample_bytree'],
                           random_state=42,
                           eval_metric='logloss',
                           tree_method='gpu_hist'
                           )

X_tr, X_val, y_tr, y_val = train_test_split(X_train, y_train, test_size=0.3, random_state=0)

In [ ]:
eval_list = [(X_tr, y_tr), (X_val, y_val)]

best_model.fit(X_tr, y_tr,
               early_stopping_rounds=100,
               eval_set=eval_list,
               eval_metric='auc'
               )

[0]	validation_0-auc:0.70170	validation_1-auc:0.69274
[1]	validation_0-auc:0.74094	validation_1-auc:0.72144
[2]	validation_0-auc:0.74818	validation_1-auc:0.72418
[3]	validation_0-auc:0.75586	validation_1-auc:0.73146
[4]	validation_0-auc:0.75838	validation_1-auc:0.73315
[5]	validation_0-auc:0.75989	validation_1-auc:0.73528
[6]	validation_0-auc:0.80785	validation_1-auc:0.78655
[7]	validation_0-auc:0.80728	validation_1-auc:0.78554
[8]	validation_0-auc:0.80845	validation_1-auc:0.78652
[9]	validation_0-auc:0.81935	validation_1-auc:0.79860
[10]	validation_0-auc:0.81766	validation_1-auc:0.79676
[11]	validation_0-auc:0.82507	validation_1-auc:0.80475
[12]	validation_0-auc:0.82458	validation_1-auc:0.80261
[13]	validation_0-auc:0.83249	validation_1-auc:0.81256
[14]	validation_0-auc:0.83041	validation_1-auc:0.80931
[15]	validation_0-auc:0.82875	validation_1-auc:0.80656
[16]	validation_0-auc:0.83599	validation_1-auc:0.81468
[17]	validation_0-auc:0.83712	validation_1-auc:0.81400
[18]	validation_0-au

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.5325981930337212,
              enable_categorical=False, eval_metric='logloss', gamma=0,
              gpu_id=-1, importance_type=None, interaction_constraints='',
              learning_rate=0.17220434779843125, max_delta_step=0, max_depth=4,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=140, n_jobs=12, num_parallel_tree=1,
              predictor='auto', random_state=42, reg_alpha=0, reg_lambda=1,
              scale_pos_weight=1, subsample=0.7272220988660021,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [ ]:
from sklearn.metrics import roc_auc_score
xgb_roc_score = roc_auc_score(y_test, best_model.predict_proba(X_test)[:,1])
xgb_roc_score

0.8413232972985847

# 최적화 파라미터 기준 모델 평가

n_estimator=500
learning_rate = 0.15
max_depth = 5


# LGBM 추가
%%time
<!--
from xgboost import XGBClassifier
from sklearn.metrics import roc_auc_score

xgb_clf = XGBClassifier(n_estimator=500, learning_rate=0.05, random_state=156)
xgb_clf.fit(X_tr, y_tr, early_stopping_rounds=100, eval_metric='auc', eval_set=[(X_tr, y_tr), (X_val, y_val)])
xgb_roc_score = roc_auc_score(y_test, xgb_clf.predict_proba(X_test)[:, 1])
xgb_roc_score -->

In [21]:
import lightgbm
lightgbm.__version__

'4.5.0'

{'colsample_bytree': 0.5132064441508775,
 'learning_rate': 0.056778045073597835,
 'max_depth': 8.0,
 'n_estimators': 220.0,
 'subsample': 0.6525052716600095}

In [23]:
lgbm_clf = SklearnCompatibleLGBClassifier(n_estimators=220, early_stopping_rounds=50)

In [24]:
eval_set = [(X_tr, y_tr), (X_val, y_val)]


In [27]:
lgbm_clf.fit(X_tr, y_tr, eval_set=eval_set, eval_metric='auc')

[LightGBM] [Warning] early_stopping_round is set=50, early_stopping_rounds=50 will be ignored. Current value: early_stopping_round=50
[LightGBM] [Info] Number of positive: 1658, number of negative: 40913
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.068278 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 13308
[LightGBM] [Info] Number of data points in the train set: 42571, number of used features: 242
[LightGBM] [Warning] early_stopping_round is set=50, early_stopping_rounds=50 will be ignored. Current value: early_stopping_round=50
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.038947 -> initscore=-3.205836
[LightGBM] [Info] Start training from score -3.205836
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[42]	training's auc: 0.91059	training's binary_logloss: 0.112183	valid_1's au

LGBMClassifier(early_stopping_rounds=50, n_estimators=220)

In [28]:
preds = lgbm_clf.predict(X_test)
preds

array([0, 0, 0, ..., 0, 0, 0])

In [29]:
lgbm_clf.predict_proba(X_test)[:,1]

array([0.00269141, 0.01391511, 0.00680896, ..., 0.11336295, 0.01466084,
       0.01081332])

In [31]:
from sklearn.metrics import roc_auc_score
roc_auc_score(y_test, lgbm_clf.predict_proba(X_test)[:,1])

0.8384286453518205

# LGBM 하이퍼오피티 튜닝

In [41]:
from sklearn.base import BaseEstimator, ClassifierMixin
from lightgbm import LGBMClassifier

# Sklearn 호환 XGBoost 클래스 정의
class SklearnCompatibleLGBClassifier(LGBMClassifier):
    def __sklearn_tags__(self):
        return {"non_deterministic": True}

In [32]:
from hyperopt.pyll.base import scope
import numpy as np

# 하이퍼파라미터 검색공간
search_space = {  'num_leaves': hp.quniform('num_leaves', 32, 64, 1),
                  'max_depth': hp.quniform('max_depth', 100, 160, 1),
                  'learning_rate': hp.loguniform('learning_rate', np.log(0.01), np.log(0.3)),
                  'subsample':hp.uniform('subsample', 0.5, 1.0),
                  'min_child_samples': hp.quniform('min_child_samples', 60, 100, 1),
                }

In [53]:
from hyperopt import fmin, tpe, Trials, STATUS_OK
from xgboost import XGBClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold

def objective_func_lgbm(params):

    model = SklearnCompatibleLGBClassifier(
        n_estimators= 100, #int(params['n_estimators']),
        num_leaves=int(params['num_leaves']),
        max_depth=int(params['max_depth']),
        learning_rate=params['learning_rate'],
        #subsample=params['subsample'],
        min_child_samples=int(params['min_child_samples']),

        random_state=42,
        eval_metric='logloss',
        early_stopping_rounds=30,
    )
    #k-fold 추가
    roc_auc_list = []

    kf = KFold(n_splits=3, shuffle=True, random_state=42)
    for tr_idx, val_idx in kf.split(X_train, y_train):
        tr_x, val_x = X_train.iloc[tr_idx], X_train.iloc[val_idx]
        tr_y, val_y = y_train.iloc[tr_idx], y_train.iloc[val_idx]
        model.fit(tr_x, tr_y, eval_set=[(tr_x, tr_y), (val_x, val_y)])

        score = roc_auc_score(val_y, model.predict_proba(val_x)[:,1])
        roc_auc_list.append(score)

    #score_mean = cross_val_score(model, X_train, y_train, cv=5, scoring='accuracy').mean()
    return {'loss': -1*np.mean(roc_auc_list) , 'status': STATUS_OK}


In [ ]:
from hyperopt import fmin, tpe, Trials, STATUS_OK
trials = Trials()
best = fmin(fn=objective_func_lgbm, space=search_space, algo=tpe.suggest
            , max_evals=50
            , trials=trials
            , rstate=np.random.default_rng(42))

[LightGBM] [Warning] Unknown parameter: eval_metric
[LightGBM] [Warning] early_stopping_round is set=30, early_stopping_rounds=30 will be ignored. Current value: early_stopping_round=30
[LightGBM] [Warning] Unknown parameter: eval_metric
[LightGBM] [Info] Number of positive: 1601, number of negative: 38943
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.089563 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 12860
[LightGBM] [Info] Number of data points in the train set: 40544, number of used features: 192
[LightGBM] [Warning] Unknown parameter: eval_metric
[LightGBM] [Warning] early_stopping_round is set=30, early_stopping_rounds=30 will be ignored. Current value: early_stopping_round=30
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.039488 -> initscore=-3.191471
[LightGBM] [Info] Start training from score -3.191471
Training until va